File: src/main/java/com/warren/loadouts/ItemLoadout.java
package com.warren.loadouts;

import java.util.ArrayList;
import java.util.List;

import org.powbot.api.rt4.Equipment;
import org.powbot.api.rt4.Item;

public interface ItemLoadout extends Iterable<ItemEntry> {
  public default List<ItemEntry> getMissing(Iterable<? extends Item> items) {
    var results = new ArrayList<ItemEntry>();
    for (var entry : this) {
      if (!entry.contained(items))
        results.add(entry);
    }

    return results;
  }

  public default List<Item> getInvalid(Iterable<? extends Item> items) {
    var results = new ArrayList<Item>();
    outer: for (var item : items) {
      for (var entry : this) {
        if (entry.matches(item)) {
          continue outer; // item is part of the loadout
        }
      }
      
      results.add(item);
    }

    return results;
  }

}

File: src/main/kotlin/com/warren/loadouts/ItemLoadout.kt
package com.warren.loadouts

import org.powbot.api.rt4.Equipment
import org.powbot.api.rt4.Inventory
import org.powbot.api.rt4.Item

interface ItemLoadout : Iterable<ItemEntry> {
    val entries: List<ItemEntry>

    override fun iterator(): Iterator<ItemEntry> = entries.iterator()

    fun getMissing(): List<ItemEntry> {
        return getMissing(Inventory.stream().list() + Equipment.stream().list())
    }

    fun getMissing(items: Iterable<Item>): List<ItemEntry> {
        val missing = ArrayList<ItemEntry>()
        for (entry in this) {
            if (!entry.contained(items)) missing += entry
        }
        return missing
    }

    fun getInvalid(items: Iterable<Item>): List<Item> {
        val invalid = ArrayList<Item>()
        outer@ for (item in items) {
            for (entry in this) {
                if (entry.matches(item)) continue@outer
            }
            invalid += item
        }
        return invalid
    }
}